alvr/graphics/resources/stream.wgsl

```diff
@@ -163,16 +163,51 @@ fn fragment_main(@location(0) uv: vec2f) -> @location(0) vec4f {
     return vec4f(color, alpha);
 }
 
-fn chroma_key_mask(color: vec3f) -> f32 {
-    let start_max = vec3f(pc.ck_channel0.x, pc.ck_channel1.x, pc.ck_channel2.x);
-    let start_min = vec3f(pc.ck_channel0.y, pc.ck_channel1.y, pc.ck_channel2.y);
-    let end_min = vec3f(pc.ck_channel0.z, pc.ck_channel1.z, pc.ck_channel2.z);
-    let end_max = vec3f(pc.ck_channel0.w, pc.ck_channel1.w, pc.ck_channel2.w);
-
-    let start_mask = smoothstep(start_min, start_max, color);
-    let end_mask = smoothstep(end_min, end_max, color);
+fn ring_trapezoid(x: f32, t1: f32, t2: f32, t3: f32) -> f32 {
+    // 0        t3--(0)
+    //  \      /
+    //   t1--t2
+    return min(smoothstep(0, -t1, -x), smoothstep(t2, t3, x));
+}
 
-    return max(start_mask.x, max(start_mask.y, max(start_mask.z, max(end_mask.x, max(end_mask.y, end_mask.z)))));
+fn chroma_key_mask(color: vec3f) -> f32 {
+    var start_max = vec3f(pc.ck_channel0.x, pc.ck_channel1.x, pc.ck_channel2.x);
+    var start_min = vec3f(pc.ck_channel0.y, pc.ck_channel1.y, pc.ck_channel2.y);
+    var end_min = vec3f(pc.ck_channel0.z, pc.ck_channel1.z, pc.ck_channel2.z);
+    var end_max = vec3f(pc.ck_channel0.w, pc.ck_channel1.w, pc.ck_channel2.w);
+
+    let sv_value = max(
+        max(
+            smoothstep(start_min.y, start_max.y, color.y),
+            smoothstep(start_min.z, start_max.z, color.z)
+        ),
+        1 - min(
+            smoothstep(end_min.y, end_max.y, color.y),
+            smoothstep(end_min.z, end_max.z, color.z)
+        )
+    );
+    var temp = 0.0;
+    // assure hue is ordered between outer bounds
+    if end_max.x < start_max.x {
+        temp = end_max.x;
+        end_max.x = start_max.x;
+        start_max.x = temp;
+    }
+    let x = fract(color.x - start_max.x);
+    var t1 = fract(start_min.x - start_max.x);
+    var t2 = fract(end_min.x - start_max.x);
+    // assure hue is ordered between inner bounds
+    if t2 < t1 {
+       temp = t1;
+       t1 = t2;
+       t2 = temp;
+    }
+    let hue_range = end_max.x - start_max.x;
+    // check hue_range, and assure inner bounds contained in outer
+    if hue_range > 1 || t1 > hue_range || t2 > hue_range {
+        return sv_value;
+    }
```

Comment:
I'm thinking, why not having a separate UI switch to control if to ignore hue selection? This is way more clear. Then to the shader we can pass this flag, or maybe use a sentinel value, for example making end_max way bigger.

The second and third condition are also not needed if the range values are correctly ordered

Reply:
all good observations

Reply:
I've removed the second and third checks.